bottlecap/src/traces/trace_processor.rs

```diff
@@ -0,0 +1,329 @@
+use datadog_trace_utils::trace_utils::SendData;
+use datadog_trace_utils::trace_utils::{self};
+
+use super::trace_agent::{ApiVersion, MAX_CONTENT_LENGTH};
+
+#[async_trait]
+pub trait TraceProcessor {
+    /// Deserializes traces from a hyper request body and sends them through the provided tokio mpsc
+    /// Sender.
+    async fn process_traces(
+        &self,
+        config: Arc<config::Config>,
+        req: Request<Body>,
+        tx: Sender<trace_utils::SendData>,
+        tags_provider: Arc<provider::Provider>,
+        version: ApiVersion,
+    ) -> http::Result<Response<Body>>;
+}
+#[derive(Clone)]
+#[allow(clippy::module_name_repetitions)]
+pub struct ServerlessTraceProcessor {
+    pub obfuscation_config: Arc<obfuscation_config::ObfuscationConfig>,
+}
+
+#[async_trait]
+impl TraceProcessor for ServerlessTraceProcessor {
+    async fn process_traces(
+        &self,
+        config: Arc<config::Config>,
+        req: Request<Body>,
+        tx: Sender<trace_utils::SendData>,
+        tags_provider: Arc<provider::Provider>,
+        version: ApiVersion,
+    ) -> http::Result<Response<Body>> {
+        info!("Recieved traces to process");
+        let (parts, body) = req.into_parts();
+
+        if let Some(response) = http_utils::verify_request_content_length(
+            &parts.headers,
+            MAX_CONTENT_LENGTH,
+            "Error processing traces",
+        ) {
+            return response;
+        }
+
+        let tracer_header_tags = (&parts.headers).into();
+
+        // deserialize traces from the request body, convert to protobuf structs (see trace-protobuf
+        // crate)
+        let (body_size, traces) = match version {
+            ApiVersion::V04 => match trace_utils::get_traces_from_request_body(body).await {
+                Ok(result) => result,
+                Err(err) => {
+                    return log_and_create_http_response(
+                        &format!("Error deserializing trace from request body: {err}"),
+                        StatusCode::INTERNAL_SERVER_ERROR,
+                    );
+                }
+            },
+            ApiVersion::V05 => match trace_utils::get_v05_traces_from_request_body(body).await {
+                Ok(result) => result,
+                Err(err) => {
+                    return log_and_create_http_response(
+                        &format!("Error deserializing trace from request body: {err}"),
+                        StatusCode::INTERNAL_SERVER_ERROR,
+                    );
+                }
+            },
+        };
+
+        let payload = trace_utils::collect_trace_chunks(
+            traces,
+            &tracer_header_tags,
+            |chunk, _root_span_index| {
+                // trace_utils::set_serverless_root_span_tags(
+                //     &mut chunk.spans[root_span_index],
+                //     config.function_name.clone(),
+                //     &config.env_type,
+                // );
```

Comment:
Should we remove this or keep it?

Reply:
Good question, as far as I can tell we need these tags on every span, not just the root span. We can discuss this on slack a bit.